### PR TITLE
Fix 8077

### LINKED ISSUE #8077: v3.4 - Diff window no longer compares the head of 2 selected branches
Up to version v3.3.1, when I selected 2 commits (branches), I got an actual diff of those 2 commits (cumulative), but in 3.4, this no longer works.

3.3.1:

image

3.4:

image

### PR DESCRIPTION
**Pull request by RussKie**
Fixes #8077

## Proposed changes

- Add "no items" to empty groups
- Collapse empty diff groups

## Screenshots 

image

New screenshots:
* image
* image

## Test methodology 

- manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
The code is "Sunday midnight" quality (i.e. far from perfect). Right now more to gather feedback and form the direction.

**Comment by RussKie** [comment]
/cc: @laurin1 your feedback is welcome too

**Comment by laurin1** [comment]
> /cc: @laurin1 your feedback is welcome too

Not sure if you wanted the reply here or in #8077. I prefer just turn all of that off (now that I know how) and as long as disabling is still an option, I don't have a strong opinion about how that info is handled with the secondary diff's enabled. 

However, I do want to thank you for your quick and thorough response and work being done on this issue - I was a little frustrated when I wrote my initial comments and your professionalism is much appreciated.

**Comment by RussKie** [comment]
> Suggest "No items" is removed, otherwise what me and @mstv discussed - I could not implement the no-collapsing.

Apologies, I don't follow.
It is impossible to show empty groups, so a group has to have at least one entry. Hence we need "no items" (or whatever other wording we choose).

> To add such an item affects a lot of parts in the UI (context menus etc). #7978 added such a hack for error messages, but this is a more common case.

Again, apologies, I can't see a direct connection of this issue to #7978. 
If you refer to a question of adding a dummy item to show otherwise hidden from use information - then yes, I think we should do so (where logical).

**Comment by gerhardol** [comment]
> It is impossible to show empty groups, so a group has to have at least one entry. Hence we need "no items" (or whatever other wording we choose).

OK, then there is no other way to show the group than adding a dummy

> Again, apologies, I can't see a direct connection of this issue to #7978.
If you refer to a question of adding a dummy item to show otherwise hidden from use information - then yes, I think we should do so (where logical).

#7978 added the "error" item to the view, this is also a dummy GitStatusItem

But then this item need special handling.

File: GitUI/UserControls/FileStatusList.cs [hunk 1]
```diff
@@ -970,7 +973,29 @@ private void UpdateFileStatusListView(bool updateCausedByFilter = false)
...
+                if (isFirstGroup && i.statuses.Count < 1)
+                {
+                    itemStatuses = noItemStatuses;
+                }
+                else
+                {
+                    itemStatuses = i.statuses;
+                }
+
+                if (GitItemStatusesWithDescription.Count > 1)
```
Comment (mstv) [inline GitUI/UserControls/FileStatusList.cs:985]: Why not `> 0`?
```suggestion
                if (GitItemStatusesWithDescription.Count > 0)
```

Comment (gerhardol) [reply GitUI/UserControls/FileStatusList.cs:985]: (Not confirmed this, but the ListGroup is Collapsible by default so this is not needed. But the code would be simpler if the compare and set state is removed.)

**Comment by RussKie** [comment]
> Why are we trying to tweak the `ListView`? Why not switch to a `TreeView` [as discussed](https://github.com/gitextensions/gitextensions/issues/1326#issuecomment-602500732)? After 3.4?

This issue is for v3.4. I too find the current behaviour highly confusing. 

> For 3.4, collapsing without the dummy item "no items" might suffice

A group isn't shown _unless_ it has items, so it is imperative to have a dummy item.

File: GitUI/UserControls/FileStatusList.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/UserControls/FileStatusList.cs:985]: Yes, can be 0. Thank you

**Comment by gerhardol** [comment]
>> For 3.4, collapsing without the dummy item "no items" might suffice

> A group isn't shown unless it has items, so it is imperative to have a dummy item.

Yes, but you still see no expanded groups, so there is a visual cue: No diff shown

https://github.com/gitextensions/gitextensions/issues/8077#issuecomment-626961077
> (Collapsing the secondary diffs by default is a good compromise IMO and will give a clear indication. I must use it a while, until I can say whether I personally like the collapsed secondary diffs or whether it annoys me to need to double click them.)

To expand secondary groups by default is a little annoying to me, the secondary groups are as important as the first when comparing branches.
Saved somehow by that empty groups are not shown

**Comment by RussKie** [comment]
Ready for reviews

**Comment by gerhardol** [comment]
This solution is mostly OK in normal use, also for myself
A limitation is that the guid is hardcoded and contents of a file "No Changes" would be shown
I propose a different approach, see https://github.com/gerhardol/gitextensions/pull/new/bugfix/_fix_8077

File: GitUI/UserControls/FileStatusList.cs [hunk 2]
```diff
@@ -1282,6 +1326,25 @@ private void FileStatusListView_ContextMenu_Opening(object sender, CancelEventAr
                 cm.Items.Add(new ToolStripSeparator());
                 cm.Items.Add(_sortByContextMenu);
             }
+
+            if (!cm.Items.Find(ShowAllDiferencesItemName, true).Any())
```
Comment (gerhardol) [inline GitUI/UserControls/FileStatusList.cs:1330]: This menu appears also in other forms, can be confusing there
The tooltip is good here, but I am leaning on removing the option anyway

Comment (RussKie) [reply GitUI/UserControls/FileStatusList.cs:1330]: A little more information please

**Comment by RussKie** [comment]
> propose a different approach, see https://github.com/gerhardol/gitextensions/pull/new/bugfix/_fix_8077

I like that. feel free to push in this branch

**Comment by gerhardol** [comment]
> feel free to push in this branch

I rebased your changes on #8114 (#8101 for 3.4, we need that too, definitely for testing this) and squashed with a new commit message.
Added two fixup commits:
 * Use GitItemStatus IsNonFile to handle the special files. Somethings I am not too happy about:
   - IsNonFile should maybe be renamed. IsStatusOnly or IsNonGit is maybe better but seem strange in GitItemStatus...
   - ErrorMessage could be used as before, handling non null as non GitStatusItem. No Changes would then be an empty message, most changes in this commit reverted 
* Only show the toggle for changing the settings if relevant. I probably want to remove this still

Of course feel free to reverse.

Will probably approve this after a little thinking. This is better than the current functionality.

File: GitUI/GitUIExtensions.cs [hunk 3]
```diff
@@ -58,7 +58,7 @@ public static class GitUIExtensions
             {
                 if (!string.IsNullOrWhiteSpace(defaultText))
                 {
-                    return fileViewer.ViewTextAsync(item?.Item?.Name, defaultText);
+                    return fileViewer.ViewTextAsync(item.Item.Name, defaultText);
```
Comment (RussKie) [inline GitUI/GitUIExtensions.cs:61]: @gerhardol this may cause NRE ❗

**Comment by RussKie** [comment]
> * IsNonFile should maybe be renamed. IsStatusOnly or IsNonGit is maybe better but seem strange in GitItemStatus...

I'm happy with `IsStatusOnly`. `IsNonGitStatus`?

File: GitUI/GitUIExtensions.cs (hunk 3 above)
Comment (gerhardol) [reply GitUI/GitUIExtensions.cs:61]: Checked on line 57, not needed

**Comment by gerhardol** [comment]
> I'm happy with IsStatusOnly

Unless someone objects, I will change to IsStatusOnly

File: GitUI/GitUIExtensions.cs (hunk 3 above)
Comment (RussKie) [reply GitUI/GitUIExtensions.cs:61]: Here's the logic lines:57-61:

```
If `item == null OR item.Item == null OR item.SecondRevision == null OR item.SecondRevision.ObjectId == null` Then
  If `!string.IsNullOrWhiteSpace(defaultText)` Then
    return fileViewer.ViewTextAsync(item.Item.Name, defaultText);
```

So `item.Item.Name` can be `item == null OR item.Item == null` and thus result in NRE.

**Comment by RussKie** [comment]
> Unless someone objects, I will change to IsStatusOnly

👍 
Btw I have rebased, so fetch the latest.

**Comment by RussKie** [comment]
I did a little tweak:
image

**Comment by mstv** [comment]
The tooltip comes and goes at the bottom of the screen (I am not requesting a change):

de9b1a46-843d-4ff1-945f-cc457b8af438

**Comment by RussKie** [comment]
> The tooltip comes and goes at the bottom of the screen (I am not requesting a change):

That's likely a Windows Forms / Win32 behaviour

**Comment by RussKie** [comment]
Tests need updating

**Comment by gerhardol** [comment]
Tests are fixed in #8119 (as well as a minor problem in addition to the core feature).
(that also rebases this PR on master)

I prefer to not backport the fix and suggest we either include #8119 in 3.4 or wait with the correction.

File: GitUI/UserControls/FileStatusList.cs [hunk 4]
```diff
@@ -976,17 +999,29 @@ private void UpdateFileStatusListView(bool updateCausedByFilter = false)
                     FileStatusListView.Groups.Add(group);
                 }
 
-                foreach (var item in i.statuses)
+                if (hasChanges && i.statuses.Count < 1)
```
Comment (mstv) [inline GitUI/UserControls/FileStatusList.cs:1002]: Move the variable to the inner block and improve the readability of the condition:
```suggestion
                IReadOnlyList<GitItemStatus> itemStatuses;
                if (hasChanges && i.statuses.Count == 0)
```

File: GitUI/UserControls/FileStatusList.cs [hunk 5]
```diff
@@ -91,6 +99,15 @@ public FileStatusList()
 
             _fullPathResolver = new FullPathResolver(() => Module.WorkingDir);
             _revisionTester = new GitRevisionTester(_fullPathResolver);
+            _noItemStatuses = new[]
+            {
+                new GitItemStatus
+                {
+                    Name = $"     - {NoFiles.Text} -",
+                    IsStatusOnly = true,
+                    ErrorMessage = ""
```
Comment (mstv) [inline GitUI/UserControls/FileStatusList.cs:108]: `ErrorMessage` should be assigned either `null`, i.e. not assigned, or `string.Empty`, shouldn't it?

Comment (gerhardol) [reply GitUI/UserControls/FileStatusList.cs:108]: Empty string has a special meaning: No icon in the status list.

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 6]
```diff
@@ -1114,13 +1114,11 @@ private void LoadUnstagedOutput(IReadOnlyList<GitItemStatus> allChangedFiles)
 
             foreach (var fileStatus in allChangedFiles)
             {
-                if (!string.IsNullOrWhiteSpace(fileStatus.ErrorMessage))
+                if (fileStatus.IsStatusOnly)
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormCommit.cs:1117]: A motivation to have these separate is that this is really two different pieces of information, but I am adding a comment too

**Pull request merged**